Safewaan/HonkedIn
# Add functionality for users to edit and cancel their events

## Trello ticket link

[As a user, I want to be able to edit the displayed information for my events.](https://trello.com/c/M4fvxucS/10-as-a-user-i-want-to-be-able-to-edit-the-displayed-information-for-my-events)
[As a user, I want to be able to cancel any of the events I created so that other users are informed that the event is no longer happening.](https://trello.com/c/xFV2GcBA/27-as-a-user-i-want-to-be-able-to-cancel-any-of-the-events-i-created-so-that-other-users-are-informed-that-the-event-is-no-longer-h)


## Implementation description
* Create `cancelEvent` API to cancel events
* Create `editEvent` API to edit events
* Create `getEventsByUser` API to search for events created by a specified user
* Create `myEvents` page which displays events made by the user logged in. The user can edit or cancel the event only if it is an active event. 



## Steps to test
Page Tests:
1. Validate that the 'my-events' endpoint brings up the "My Events" page.
2. On the `myEvents` page, ensure that the user's created events appear.

Edit Event Tests:
1. Click the "Edit Event" button. Verify that a form appears, with all the respective fields. Validate that the fields are pre-filled with the event's current data.
2. Validate that pressing "Close" closes the form and no changes are made. 
3. Set all the fields to empty (except date, date cannot be set to empty). Validate that an error message appears for each field if it is left empty, and "Edit Event" is pressed.
4. Input valid data for the fields, and press the "Edit Event" button. Validate that a success toast message appears, and the page refreshes after a short delay. Verify that the event is updated with any changes made in the form.
5. Verify that the changes are reflected in the db. 

Cancel Event Tests:
1. Click the "Cancel Event" button. Verify that a confirmation message appears. Validate that pressing "no" does nothing. Validate that pressing "yes" displays a success toast message, and refreshes the page after a short delay. Validate that the event's status is set to "Cancelled". 
2. Validate that the cancelled event does not have the option for editing, or cancelling.
3. Validate that the matching event is set to "Cancelled" in the db.
4. Navigate back to the "Events" page. Validate that the event can no longer be joined by other users.



## What should reviewers focus on?
* There is a ton of reused code, especially for the edit feature. Please test all error cases and ensure exceptions are handled correctly.
* Currently, once an event is cancelled, it can no longer be reactivated. Not sure if we want to change this. Also unsure of the third status, "Completed". We can set up a query to run at the end of every day to update all events that have passed (by doing a date check) but this may be too much work for a small feature. We can discuss this if needed.
